src/java/org/apache/cassandra/schema/PartitionDenylist.java

```
@@ -0,0 +1,406 @@
+               !SchemaConstants.TRACE_KEYSPACE_NAME.equals(keyspace) &&
+               !SchemaConstants.AUTH_KEYSPACE_NAME.equals(keyspace);
+    }
+
+    private boolean isPermitted(final TableId tid)
+    {
+        final TableMetadata tmd = Schema.instance.getTableMetadata(tid);
+        if (tmd == null)
+            return false;
+        return isPermitted(tmd.keyspace);
+    }
+
+    public boolean validateKey(final String keyspace, final String cf, final ByteBuffer key)
+    {
+        return validateKey(getTableId(keyspace, cf), key);
+    }
+
+    public boolean validateKey(final TableId tid, final ByteBuffer key)
+    {
+        if (!DatabaseDescriptor.enablePartitionDenylist() || tid == null || !isPermitted(tid))
+            return true;
+
+        try
+        {
+            return !denylist.get(tid).keys.contains(key);
+        }
+        catch (final ExecutionException e)
+        {
+            // In the event of an error accessing or populating the cache, assume it's not denylisted
+            logAccessFailure(tid, e);
+            return true;
+        }
+    }
+
+    private void logAccessFailure(final TableId tid, Throwable e)
+    {
+        final TableMetadata tmd = Schema.instance.getTableMetadata(tid);
+        if (tmd == null)
+            logger.debug("Failed to access partition denylist cache for unknown table id {}", tid.toString(), e);
+        else
+            logger.debug("Failed to access partition denylist cache for {}/{}", tmd.keyspace, tmd.name, e);
+    }
+
+    /**
+     * @return number of denylisted keys in range
+     */
+    public int validateRange(final String keyspace, final String cf, final AbstractBounds<PartitionPosition> range)
+    {
+        return validateRange(getTableId(keyspace, cf), range);
+    }
+
+    /**
+     * @return number of denylisted keys in range
+     */
+    public int validateRange(final TableId tid, final AbstractBounds<PartitionPosition> range)
```

Maybe that's just me, but naming most of the methods is confusing for me (they are verbs and I have certain expectations when I see them). For example, method `validate` should throw an exception if something is invalid. `getInvalidKeysCount` seems to be more clear for me in this case.

The same:
`validateKey` -> `isKeyValid`
`denylist` -> `addToDenylist`

---

Completely agree and took a pass at renaming things. Let me know what you think when I push it up.